hoot-services/src/main/java/hoot/services/controllers/osm/map/FolderResource.java

```diff
@@ -0,0 +1,369 @@
+        if(user != null) {
+            // public or folder owned by current user
+            sql.where(folders.publicCol.isTrue().or(folders.userId.eq(user.getId())));
+        }
+        List<FolderMapMappings> linkRecordSet = sql.orderBy(folderMapMappings.folderId.asc()).fetch();
+        linkRecords = mapLinkRecordsToLinks(linkRecordSet);
+        return linkRecords;
+    }
+
+    /**
+     * Adds new dataset folder
+     *
+     * POST hoot-services/osm/api/0.6/map/folders/add/{parent}/{name}
+     *
+     * @param folderName
+     *            Display name of folder
+     * @param parentId
+     *            The parent folder of the new folder. If at root level, is
+     *            equal to 0.
+     * @param isPublic
+     *            Default true
+     * @return jobId Success = True/False
+     */
+    @POST
+    @Path("/add/{parentId : \\d+}/{folderName}")
+    @Consumes(MediaType.TEXT_PLAIN)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response addFolder(@Context HttpServletRequest request, @PathParam("folderName") String folderName, @PathParam("parentId") Long parentId, @DefaultValue("true") @QueryParam("isPublic") Boolean isPublic) {
+        Users user = Users.fromRequest(request);
+        Long userid = -1L;
+        if(user != null) {
+            userid = user.getId();
+        }
+
+        // handles some ACL logic for us...
+        // (to be sure we can see the parent folder)
+        getFolderForUser(user, parentId);
+
+
+        Long newId = createQuery()
+                .select(Expressions.numberTemplate(Long.class, "nextval('folders_id_seq')"))
+                .from()
+                .fetchOne();
+
+        Timestamp now = new Timestamp(System.currentTimeMillis());
+
+
+        createQuery()
+                .insert(folders).columns(folders.id, folders.createdAt, folders.displayName, folders.publicCol, folders.userId, folders.parentId)
+                .values(newId, now, folderName, isPublic, userid, parentId).execute();
+
+        java.util.Map<String, Object> ret = new HashMap<String, Object>();
+        ret.put("success", true);
+        ret.put("folderId", newId);
+
+        return Response.ok().entity(ret).build();
+    }
+
+    /**
+     * Deletes a dataset folder.
+     *
+     * POST hoot-services/osm/api/0.6/map/folder/delete/{id}
+     *
+     *
+     * @param folderId
+     *            Folder Id
+     * @return jobId
+     */
+    @DELETE
+    @Path("/{folderId : \\d+}")
+    @Consumes(MediaType.TEXT_PLAIN)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response deleteFolder(@Context HttpServletRequest request, @PathParam("folderId") Long folderId) {
+        Users user = Users.fromRequest(request);
+        Folders folder = getFolderForUser(user, folderId);
+
+        createQuery()
+            .update(folders)
+            .where(folders.parentId.eq(folderId))
+            .set(folders.parentId, folder.getParentId()) // move stuff in this folder to -this folder's- parent.
```

Comment:
seems like deleting a folder deletes all folders and datasets beneath it in my testing, so this commend confuses me.

Reply:
I will add unit testing to confirm logic + bolster confidence

Reply:
This does not occur currently, the UI is clearing folders for us w/ successive deletes